### ISSUE TITLE
No database contention in case of partial unique index

### ISSUE THREAD

**Issue by SButterfly** [cockroachdb/cockroach#112856]
**Describe the problem**
Hi!
I found an issue with a missing database contention if we update to row in parallel affecting one partial unique index.
I think that this **is a bug**, because I have compared the behaviour with PostgreSQL (serializable level), and **PostgreSQL is throwing 'could not serialize access due to read/write dependencies among transactions'.**

**To Reproduce**

Let me explain the business problem first, cause it would be easier to understand how did we get the exception.
Let's imagine, that we want to have a table with unique users, but with option to soft delete:
```sql
create table users
(
    -- Unique id of the user (internal id)
    id         uuid primary key,
    -- unique id of the user (external id) (excluding deleted, see index)
    code       uuid not null,
    -- user name
    name       text not null,
    -- soft delete flag
    deleted_at timestamptz
);

CREATE UNIQUE INDEX idx_users_code ON users (code) WHERE deleted_at is null;
```

Let's imagine we want to soft delete user + recreate the broken user simultaneously.
```sql
begin;
var user = select * from users where deleted_at is null and id = ?

-- business code that checks and decides to soft delete the user
begin; -- in a new connection, previous connection will wait until this new connection will be finished

update users set deleted_at = now() where id = ?

commit; -- this one works good

-- this line fails with 'duplicate key value violates unique constraint "idx_users_code"', which I don't expect
-- because both transactions **touched the same value in unique index**
-- 
-- In PostgreSQL with the serializable level, this line fails with 40001 'Could not serialize access due to read/write dependencies among transactions' stand he whole transaction successfully restarted
insert into users (id, code, name, deleted_at) values(random(), user.code, user.name, null);

-- never reached in cockroachDB
-- reached in PostgreSQL, but after restart - which is expected and good
commit;
```

Cockroach: v23.1.11
Postgresql: 16

**Expected behavior**

40001 is returned when we try to insert a new row after changing the first one.

**Actual behavior**

'ERROR: duplicate key value violates unique constraint "idx_users_code"'

**Environment:**
 - CockroachDB version v23.1.11

**Additional information**
- Created simple java project with test containers to reproduce the issue: https://github.com/SButterfly/cockroachdb-missing-db-contention

Jira issue: CRDB-32644

**Comment by mgartner** [comment]
I can reproduce this in v23.1.13, but not in v23.2.1. So I think recent improvements have fixed this. @michae2 does that sound plausible to you?

**Comment by mgartner** [comment]
Friendly ping @michae2

**Comment by mgartner** [comment]
We think this is fixed as of v23.2. Please open a new issue if you find otherwise. Thanks!
